Core: Increase address space limits and rework Windows address space initialization.

StevenMiller123 opened the pull request:
Current Windows builds of shadPS4 are prone to asserting in address_space on emulator initialization due to Windows' address space randomization. This also prevents expanding the address space to higher addresses (which is needed for a handful of EA titles).

This PR introduces logic to detect and map around system mappings on Windows devices, pass information about these gaps to our memory code, and more robust address validation logic to ensure gaps are accounted for when detecting if requested memory areas are within the vma map.

For Mac users: this PR should make asserts trigger more consistently when games try mapping to addresses that we can't reserve.

For Windows users: this PR should fix those random-ish cases of
```
[Debug] <Critical> address_space.cpp:105 operator(): Assertion Failed!
Unable to reserve virtual address space: Not enough memory resources are available to process this command.
```

For all users: this PR should address cases of the following assert triggering on addresses greater than 0x11000000000 (mainly seen in earlier titles published by EA).

```
[Debug] memory.cpp:902 operator(): Assertion Failed!
Input address .... is out of bounds
```
I'm opening this as a draft for now since I need to test edge cases further, and ideally make sure our own HLE mappings aren't going to hit asserts in the more extreme cases.

I would also appreciate some tests from the aforementioned EA games, the only thing I have hitting that out of bounds assert is an app, and it dies from a Windows-specific issue with code patches.

Commit:
SearchFree adjustments

Commit:
Robust address validation

I've adjusted IsValidAddress to take in a size, and check whether the whole range is contained in vma map.
If no size is provided, the function reverts to the old form of address validation instead.

Commit:
Map around gaps

As is, this should work mostly.
Only remaining issue is adding logic to pass the "mapped regions" to the guest vma map (and make such logic cross-platform).

Commit:
Initialize vma_map using gaps

This should allow memory code to catch any issues from address space gaps, and prevent non-fixed mappings from jumping to a location that isn't actually available.

Commit:
Clang

Commit:
Fix compile

Commit:
Clang

Commit:
Fix compile again

Commit:
Set system_managed_base and system_managed_size based on

Many places in our code use system_managed_base as the minimum mappable address, ensure this fact remains the same  on Windows to prevent potential bugs.

Commit:
Reduce address validation in SearchFree

Allows SearchFree to function when a certain Windows GPU driver goes and reserves the whole system managed area.

Since SearchFree is only called on flexible addresses, allowing this particular case, where addresses are in bounds, but there's not enough space to map, should be safe enough.

Commit:
Bump address space size further

To handle Madden NFL 16 (and any games like it)

Commit:
More thorough logging of available memory regions

Should help with spotting weirdness.

Exhigh commented on src/core/memory.cpp:
```diff
@@ -138,9 +132,9 @@ void MemoryManager::CopySparseMemory(VAddr virtual_addr, u8* dest, u64 size) {
-    ASSERT_MSG(IsValidAddress(address), "Attempted to access invalid address {}",
-               fmt::ptr(address));
     const VAddr virtual_addr = std::bit_cast<VAddr>(address);
+    ASSERT_MSG(IsValidMapping(virtual_addr, num_bytes), "Attempted to access invalid address {}",
```
```suggestion
    ASSERT_MSG(IsValidMapping(virtual_addr, num_bytes), "Attempted to access invalid address {:#x}",
```

StevenMiller123 committed:
Formatting fixes

StevenMiller123 committed:
Clang

StevenMiller123 committed:
Slight reduction of user space

Still large enough to handle EA's shenanigans, but small enough that Linux doesn't break.

Commit:
Assert on VirtualQuery failure

Commit:
Extra debugging information

Commit:
Further reduce user space

This will unfix most of EA's titles, but UFC will still work.
Older windows versions support the high addresses, but trying to actually use them causes significant performance issues.

Commit:
Extra debugging info

Just in case other testers still run into issues.

Commit:
Remove debug logging

Commit:
Revert user space increases

Technically this constant is still higher than before, but weird side effects of our old logic resulted in a max address somewhere around this in main.

Commit:
address_space: Support expanded virtual memory space on macOS.

Co-Authored-By: squidbus <175574877+squidbus@users.noreply.github.com>

Commit:
Move address space constants to address_space.cpp

This ensures that all code must use the calculated address space memory values instead of the constants, since the calculated values can differ based on the platform.

This does require slight modification to thread state and gnmdriver code, since both were already using these constants directly.

Commit:
Workaround Windows 10 limitations

If a Windows 10 device is detected, use a lower value for USER_MAX to prevent system-wide hangs in VirtualAlloc2 calls.

Commit:
Fix compile for Windows-Qt

squidbus commented on src/core/libraries/gnmdriver/gnmdriver.cpp:
```diff
@@ -70,7 +71,12 @@ static bool send_init_packet{true}; // initialize HW state before first game's s
+auto* memory = Core::Memory::Instance();
+auto& address_space = memory->GetAddressSpace();
+static VAddr tessellation_factors_ring_addr = address_space.SystemReservedVirtualBase() +
+                                              address_space.SystemReservedVirtualSize() - 0xFFFFFFF;
```
Not sure if it's safe to initialize this statically, may be safer to initialize this on first `sceGnmGetTheTessellationFactorRingBufferBaseAddress` call.

StevenMiller123 committed:
Move tessellation_factors_ring_addr initialization to sceGnmGetTheTessellationFactorRingBufferBaseAddress

Commit:
Set image base address on Windows

This seems to work fine on Windows 11, needs testing from Windows 10 due to the previously discussed bugs.

StevenMiller123 committed:
Set Linux executable base to 0x700000000000

This allows Linux to map the full user space without any workarounds.

Co-Authored-By: Marcin Mikołajczyk <2052578+mikusp@users.noreply.github.com>

StevenMiller123 committed:
Basic formatting changes

StevenMiller123 committed:
Reduce USER_MAX on Linux

Seems like finding a reliable way to move shadPS4's position in memory is difficult, for now limit the user size so we aren't trying to overwrite ourselves.

squidbus commented on src/core/libraries/gnmdriver/gnmdriver.cpp:
```diff
@@ -996,7 +999,14 @@ int PS4_SYSV_ABI sceGnmGetShaderStatus() {
 
 VAddr PS4_SYSV_ABI sceGnmGetTheTessellationFactorRingBufferBaseAddress() {
+    auto* memory = Core::Memory::Instance();
+    auto& address_space = memory->GetAddressSpace();
```
Can move these into the `if` case.

squidbus reviewed (commented):
Just one small comment

Commit:
Move memory and address_space variables.
